src/middlewared/middlewared/plugins/mail.py

```diff
@@ -1,41 +1,324 @@
+from email.mime.text import MIMEText
+from email.utils import formatdate
+from lockfile import LockFile, LockTimeout
+
+import base64
+import errno
+import json
 import os
-import sys
+import pickle
+import smtplib
+import socket
+import syslog
+
+
+class QueueItem(object):
+
+    def __init__(self, message):
+        self.attempts = 0
+        self.message = message
+
+
+class MailQueue(object):
+
+    QUEUE_FILE = '/tmp/mail.queue'
+    MAX_ATTEMPTS = 3
+
+    def __init__(self):
+        self.queue = None
+
+    def append(self, message):
+        self.queue.append(QueueItem(message))
+
+    @classmethod
+    def is_empty(cls):
+        if not os.path.exists(cls.QUEUE_FILE):
+            return True
+        try:
+            return os.stat(cls.QUEUE_FILE).st_size == 0
+        except OSError:
+            return True
+
+    def _get_queue(self):
+        try:
+            with open(self.QUEUE_FILE, 'rb') as f:
+                self.queue = pickle.loads(f.read())
+        except (pickle.PickleError, EOFError):
+            self.queue = []
 
-if '/usr/local/www' not in sys.path:
-    sys.path.append('/usr/local/www')
-os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'freenasUI.settings')
+    def __enter__(self):
+        self._lock = LockFile(self.QUEUE_FILE)
+        while not self._lock.i_am_locking():
+            try:
+                self._lock.acquire(timeout=330)
+            except LockTimeout:
+                self._lock.break_lock()
 
-import django
-from django.apps import apps
-if not apps.ready:
-    django.setup()
+        if not os.path.exists(self.QUEUE_FILE):
+            open(self.QUEUE_FILE, 'a').close()
 
-from freenasUI.common.system import send_mail
+        self._get_queue()
+        return self
+
+    def __exit__(self, typ, value, traceback):
+
+        with open(self.QUEUE_FILE, 'wb+') as f:
+            if self.queue:
+                f.write(pickle.dumps(self.queue))
+
+        self._lock.release()
+        if typ is not None:
+            raise
 
 
 class MailService(ConfigService):
 
     class Config:
         datastore = 'system.email'
+        datastore_prefix = 'em_'
+        datastore_extend = 'mail.mail_extend'
+
+    @private
+    async def mail_extend(self, cfg):
+        if cfg['security']:
+            cfg['security'] = cfg['security'].upper()
+        return cfg
+
+    @accepts(Dict(
+        'mail',
+        Str('fromemail'),
+        Str('outgoingserver'),
+        Int('port'),
+        Str('security', enum=['PLAIN', 'SSL', 'TLS']),
+        Bool('smtp'),
+        Str('user'),
+        Str('pass'),
+    ))
+    async def do_update(self, data):
+        config = await self.config()
+
+        new = config.copy()
+        new.update(data)
+        new['security'] = new['security'].lower()  # Django Model compatibility
+
+        verrors = ValidationErrors()
+
+        if new['smtp'] and new['user'] == '':
+            verrors.add('user', 'This field is required when SMTP authentication is enabled')
```

Comment:
Did it work for you? I had to do `verrors.add("mail.user")`, because validator for, e.g., `security`, will raise ValidationErrors for `mail.security`

Reply:
https://i.imgur.com/iVevKJs.png

Reply:
However you're right, we should identify which schema we are referring to.